### PR TITLE
Alternative groups implementation

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Tests pass?   | ?
| Fixed issues  | #5 See https://github.com/yiisoft/router-fastroute/pull/9

### CODE REVIEW

File: src/Route.php [hunk 1]
```diff
@@ -264,8 +290,35 @@ public function getDefaults(): array
...
 
-    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    /**
+     * @internal please use {@see dispatch()} or {@see process()}
+     * @param ServerRequestInterface $request
+     * @return ResponseInterface
+     */
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $middleware = current($this->middlewares);
```
Comment (rustamwin) [inline src/Route.php]:
```suggestion
        $middleware = \current($this->middlewares);
```

File: src/Route.php [hunk 2]
```diff
@@ -264,8 +290,35 @@ public function getDefaults(): array
...
+        next($this->middlewares);
```
Comment (rustamwin) [inline src/Route.php]:
```suggestion
        \next($this->middlewares);
```

File: src/Route.php [hunk 3]
```diff
@@ -264,8 +290,35 @@ public function getDefaults(): array
...
+
+            throw new \LogicException('Middleware stack exhausted');
+        }
+
+        return $middleware->process($request, $this);
+    }
+
+    public function dispatch(ServerRequestInterface $request): ResponseInterface
+    {
+        reset($this->middlewares);
```
Comment (rustamwin) [inline src/Route.php]:
```suggestion
        \reset($this->middlewares);
```

File: src/RouterFactory.php [hunk 4]
```diff
@@ -21,7 +21,13 @@ public function __invoke(ContainerInterface $container): RouterInterface
         /* @var $router RouterInterface */
         $router = $factory($container);
         foreach ($this->routes as $route) {
-            $router->addRoute($route);
+            if ($route instanceof Route) {
+                $router->addRoute($route);
+            } elseif (is_array($route) && count($route) === 2 && is_string($route[0]) && is_callable($route[1])) {
```
Comment (rustamwin) [inline src/RouterFactory.php]:
```suggestion
            } elseif (\is_array($route) && \count($route) === 2 && \is_string($route[0]) && \is_callable($route[1])) {
```

File: src/Middleware/Callback.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+
+namespace Yiisoft\Router\Middleware;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+final class Callback implements MiddlewareInterface
```
Comment (rustamwin) [inline src/Middleware/Callback.php]: Does this closes #11 issue?

File: src/Group.php [hunk 6]
```diff
@@ -3,51 +3,60 @@
...
+    final public function addGroup(string $prefix, callable $callback): void
     {
-        $this->routes[] = $route;
+        $this->items[] = new Group($prefix, $callback);
+    }
+
+    final public function addMiddleware($middleware): self
+    {
+        if (\is_callable($middleware)) {
+            $middleware[] = new Callback($middleware);
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/Group.php]: Multiple middleware are now supported.

File: src/Middleware/Callback.php (hunk 5 above)
Comment (samdark) [reply src/Middleware/Callback.php]: Yes.

File: src/Group.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (rustamwin) [reply src/Group.php]: But the [next](https://github.com/yiisoft/router/pull/14/files#diff-ed0c1ca52987aaa4167c07defa0643e3R39) step gives an error

**Pull request merged**